handler/core/src/main/java/io/knotx/fragments/handler/consumer/metadata/MetadataConverter.java

```
@@ -0,0 +1,118 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.knotx.fragments.handler.consumer.metadata;
+
+import io.knotx.fragments.engine.FragmentEvent;
+import io.knotx.fragments.engine.NodeMetadata;
+import io.knotx.fragments.engine.TaskMetadata;
+import io.knotx.fragments.engine.api.node.NodeType;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import org.apache.commons.lang3.StringUtils;
+
+public class MetadataConverter {
+
+  private static final String METADATA_STATUS = "_metadataStatus";
+  private static final String MISSING = "missing";
+  private static final String LOG_STATUS = "_logStatus";
+
+  private final String rootNodeId;
+  private final Map<String, NodeMetadata> nodes;
+  private final EventLogConverter eventLogConverter;
+
+  private MetadataConverter(FragmentEvent event, TaskMetadata taskMetadata) {
+    this.rootNodeId = taskMetadata.getRootNodeId();
+    this.nodes = taskMetadata.getNodesMetadata();
+    this.eventLogConverter = new EventLogConverter(event.getLog().getOperations());
+  }
+
+  public static MetadataConverter from(FragmentEvent event, TaskMetadata metadata) {
+    return new MetadataConverter(event, metadata);
+  }
+
+  public JsonObject createJson() {
+    if (StringUtils.isBlank(rootNodeId)) {
+      return new JsonObject();
+    } else {
+      return createNodeJson(rootNodeId);
+    }
+  }
+
+  private JsonObject createNodeJson(String id) {
+    return Optional.of(new JsonObject())
+        .map(json -> fillWithMetadata(json, id))
+        .map(json -> eventLogConverter.fillWithLog(json, id))
+        .map(this::fillMissingNode)
+        .orElseGet(JsonObject::new);  // Should never happen
```

If this should never happen then it should throw an exception